streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -581,98 +508,48 @@ private void initTopology() {
-     *   - commit offsets
+     * the following order must be followed:
+     *  1. first close topology to make sure all cached records in the topology are processed
+     *  2. then flush the state, send any left changelog records
+     *  3. then flush the record collector
+     *  4. then commit the record collector -- for EOS this is the synchronization barrier
+     *  5. then checkpoint the state manager -- even if we crash before this step, EOS is still guaranteed
      * </pre>
      *
      * @throws TaskMigratedException if committing offsets failed (non-EOS)
      *                               or if the task producer got fenced (EOS)
      */
-    // visible for testing
-    void suspend(final boolean clean,
-                 final boolean isZombie) {
-        // this is necessary because all partition times are reset to -1 during close
-        // we need to preserve the original partitions times before calling commit
-        final Map<TopicPartition, Long> partitionTimes = extractPartitionTimes();
+    private void suspend(final boolean clean) {
         try {
-            closeTopology(); // should we call this only on clean suspend?
-        } catch (final RuntimeException fatal) {
-            if (clean) {
-                throw fatal;
-            }
-        }
-
-        if (clean) {
-            TaskMigratedException taskMigratedException = null;
-            try {
-                commit(false, partitionTimes);
-            } finally {
-                if (eosEnabled) {
-                    stateMgr.checkpoint(activeTaskCheckpointableOffsets());
-
-                    try {
-                        recordCollector.close();
-                    } catch (final RecoverableClientException e) {
-                        taskMigratedException = new TaskMigratedException(this, e);
-                    } finally {
-                        producer = null;
-                    }
-                }
-            }
-            if (taskMigratedException != null) {
-                throw taskMigratedException;
-            }
-        } else {
-            // In the case of unclean close we still need to make sure all the stores are flushed before closing any
-            super.flushState();
+            // If the suspension is from unclean shutdown, then only need to close topology and flush state to make sure that when later
+            // closing the states, there's no records triggering any processing anymore; also swallow all caught exceptions
+            closeTopology();
 
-            if (eosEnabled) {
-                maybeAbortTransactionAndCloseRecordCollector(isZombie);
+            if (clean) {
+                commitState();
+                // whenever we have successfully committed state during suspension, it is safe to checkpoint
+                // the state as well no matter if EOS is enabled or not
+                stateMgr.checkpoint(checkpointableOffsets());
+            } else {
+                stateMgr.flush();
```

Comment:
This is part of a bug fix that we discovered and did recently: previously in unclean suspension, we don't try to commit state or checkpoint state at all; but later we realized that even under unclean suspension we still need to flush the state (with clean suspension the flush is done as part of the committing state). I know it looks a bit asymmetry and we can for example extract the flush out of `commitState`, but after I looked at it for a bit I think it's better not to extract it out since `commitState` is called by a couple different callers and the flushing should actually be in the middle of the process, but not at the beginning of the process. WDYT?

Reply:
Sounds good. Maybe adding your response as a comment would be warranted, otherwise we'd have to keep reading the source code of `commitState` to keep this straight.

Reply:
Sounds good.